src/python/pants/backend/python/rules/missing_init_test.py

```diff
@@ -0,0 +1,204 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+import json
+from typing import List
+
+import pytest
+
+from pants.backend.python.rules import missing_init
+from pants.backend.python.rules.ancestor_files import find_missing_ancestor_files
+from pants.backend.python.rules.missing_init import (
+    MissingInit,
+    MissingInitRequest,
+    MissingNonEmptyInitFiles,
+)
+from pants.core.util_rules import strip_source_roots
+from pants.engine.internals.scheduler import ExecutionError
+from pants.engine.rules import RootRule
+from pants.testutil.engine.util import Params
+from pants.testutil.option.util import create_options_bootstrapper
+from pants.testutil.test_base import TestBase
+
+
+class InjectInitTest(TestBase):
+    @classmethod
+    def rules(cls):
+        return (
+            *super().rules(),
+            find_missing_ancestor_files,
+            *missing_init.rules(),
+            *strip_source_roots.rules(),
+            RootRule(MissingInitRequest),
+        )
+
+    def assert_injected(
+        self,
+        *,
+        source_roots: List[str],
+        declared_files_stripped: bool,
+        original_declared_files: List[str],
+        original_undeclared_empty_files: List[str],
+        original_undeclared_nonempty_files: List[str],
+        expected_discovered: List[str],
+    ) -> None:
+        for f in original_undeclared_empty_files:
+            self.create_file(f, "")
+        for f in original_undeclared_nonempty_files:
+            self.create_file(f, "# nonempty")
+        request = MissingInitRequest(
+            self.make_snapshot({fp: "" for fp in original_declared_files}),
+            sources_stripped=declared_files_stripped,
+        )
+        bootstrapper = create_options_bootstrapper(
+            args=[f"--source-root-patterns={json.dumps(source_roots)}"]
+        )
+        result = self.request_single_product(MissingInit, Params(request, bootstrapper)).snapshot
+        assert list(result.files) == sorted(expected_discovered)
+
+    def test_unstripped(self) -> None:
+        self.assert_injected(
+            source_roots=["src/python", "tests/python"],
+            declared_files_stripped=False,
+            original_declared_files=[
+                "src/python/project/lib.py",
+                "src/python/project/subdir/__init__.py",
+                "src/python/project/subdir/lib.py",
+                "src/python/no_init/lib.py",
+            ],
+            original_undeclared_empty_files=[
+                "src/python/project/__init__.py",
+                "tests/python/project/__init__.py",
+            ],
+            original_undeclared_nonempty_files=[],
+            expected_discovered=["src/python/project/__init__.py"],
```

Comment:
I don't follow why `tests/python/project/__init__.py` isn't included.

Reply:
Because there are no .py files under tests/python that are included in the "declared" set from which to do the discovering.